### PR TITLE
Allow specifying more options per endpoint (e.g. access callback)

### PR DESCRIPTION
**Pull request by e0ipso**
For example:

``` php
protected $controllers = array(
  '' => array(
    // Shorthand.
    \RestfulInterface::GET => 'getList',
  ),
  '\d+' => array(
    // Verbose
    \RestfulInterface::GET => array(
      'method'=> 'viewEntity',
      'access callback' => array($this, 'accessViewEntity'),
    ),
  ),
);
```

So the access could be moved to it's own method per endpoint.

### CODE REVIEW

**Comment by e0ipso** [comment]
@amitaibu please review.

**Comment by e0ipso** [comment]
Tests, including new ones are green.

**Comment by amitaibu** [comment]
Do we want to add the access callback per public-field also in this one, or follow up?

**Comment by e0ipso** [comment]
> Do we want to add the access callback per public-field also in this one, or follow up?

Do you mean additionally to the entity property access? I'm thinking that maybe that is a bit too much and can be handled within the `callback` for the property or even `process callbacks` if it's for removing stuff instead of throwing an exception.

**Comment by amitaibu** [comment]
> if it's for removing stuff instead of throwing an exception.

access callback would be about allowing to access the property - not an exception.

Use case: In a `users` resource I'd like to have the `password` property, so client can register a new user with their password via POST. However, on GET (e.g. `api/v1/users/1`), I'd like to make sure password's access == FALSE

**Comment by e0ipso** [comment]
Will it need to be a callback? Does it suffice to have an optional array of fields to hide? Like:

``` php
array(
  \RestfulInterface::get => array(
    'callback' => 'viewEntities',
    'access callback' => 'callback_for_the_GET_method',
    'hide fields' => array('label', 'password'),
  ),
);
```

**Comment by amitaibu** [comment]
> Will it need to be a callback?

Yes, because maybe you want to deny access only to certain users, by arbitrary conditions.

Also, I think this feature makes a lot of sense, in terms of expected granularity: We have access on the resource level, this PR is for the method level, and next is for the public field level.

**Comment by e0ipso** [comment]
@amitaibu if it has to be a callback then it should be specified in the public field configuration array.

I opened a follow up for this. https://github.com/Gizra/restful/issues/180

**Comment by e0ipso** [comment]
This can be merged.

File: tests/modules/restful_test/plugins/restful/node/test_articles/1.3/RestfulTestArticlesResource__1_3.class.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+   *
+   * @return boolean
+   *   TRUE for access granted, FALSE otherwise.
+   */
+  protected function accessViewEntityFalse() {
+    return FALSE;
+  }
+
+  /**
+   * Custom access callback for the GET method.
```
Comment (amitaibu) [inline tests/modules/restful_test/plugins/restful/node/test_articles/1.3/RestfulTestArticlesResource__1_3.class.php]: GET => HEAD

File: tests/RestfulEntityAndPropertyAccessTestCase.test [hunk 2]
```diff
@@ -169,4 +169,23 @@ class RestfulEntityAndPropertyAccessTestCase extends DrupalWebTestCase {
       $this->pass('Non-privileged user cannot view entity.');
     }
   }
+
+  public function testEnpointAccessCallback() {
```
Comment (amitaibu) [inline tests/RestfulEntityAndPropertyAccessTestCase.test:176]: Can you add PHPdocs please

File: tests/RestfulEntityAndPropertyAccessTestCase.test [hunk 3]
```diff
@@ -169,4 +169,23 @@ class RestfulEntityAndPropertyAccessTestCase extends DrupalWebTestCase {
...
+    $settings = array(
+      'type' => 'article',
+    );
+    $node = $this->drupalCreateNode($settings);
+
+    $handler = restful_get_restful_handler('test_articles', 1, 3);
+    try {
+      $handler->get($node->nid);
+      $this->fail('Custom access callback per endpoint not executed correctly.');
```
Comment (amitaibu) [inline tests/RestfulEntityAndPropertyAccessTestCase.test]: Maybe `"Custom access callback per resource's method not executed correctly."` or does endpoint imply a method?

**Comment by amitaibu** [comment]
After the minor comments, this can be merged. Thanks.

**Comment by e0ipso** [comment]
Comment changes done.

**Comment by amitaibu** [comment]
Merged, thanks.

**Pull request closed without merging**